dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,193 @@ public Pattern getParameterPattern() {
+            ResultSet rs = getMetaData().getProcedureColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+            return rs;
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs) throws SQLException {
+            switch (rs.getInt("COLUMN_TYPE")) {
+                case DatabaseMetaData.procedureColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.procedureColumnInOut:
+                    return Direction.INPUT_OUTPUT;
+                case DatabaseMetaData.procedureColumnOut:
+                    return Direction.OUTPUT;
+                default:
+                    throw new SQLException("Unpexpected procedure parameter type for parameter " +
+                                  rs.getString("COLUMN_NAME"));
+            }
+        }
+    }
+
+    private class DatabaseFunction extends DatabaseObject {
+
+        private DatabaseFunction(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        public boolean exists() throws SQLException {
+            try (ResultSet rs = getObjects()) {
+                return rs.next();
+            }
+        }
+
+        @Override
+        protected ResultSet getObjects() throws SQLException {
+            return getMetaData().getFunctions(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getMetaData().getFunctionColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs) throws SQLException {
+            int paramType = rs.getInt("COLUMN_TYPE");
+            switch (paramType) {
+                case DatabaseMetaData.functionColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.functionReturn:
+                    return Direction.RETURN_VALUE;
+                default:
+                    throw new SQLException("Unpexpected function parameter type for parameter " +
+                                  rs.getString("COLUMN_NAME"));
+            }
+        }
+    }
+
+    private class DatabaseObjectName {
+        String schemaName, objectName;
+
+        private DatabaseObjectName(String defaultSchemaName, String objName) {
+            String[] qualifiers = objName.split("\\.");
+            schemaName = (qualifiers.length == 2) ? qualifiers[0] : defaultSchemaName;
+            objectName = (qualifiers.length == 2) ? qualifiers[1] : qualifiers[0];
+        }
+
+        private String getSchemaName() {
+            return schemaName;
+        }
+
+        String getObjectName() {
+            return objectName;
+        }
+    }
+
+    public Map<String, DbParameterAccessor> getAllProcedureParameters(String callableName)
+                                                                       throws SQLException {
```

Comment:
My approach for throws formatting is also not very streamlined. Usually I try to keep all indentations at position divisible by 4 spaces (or whatever the unit is). Things I might do for that case:

``` java
public Map<String, DbParameterAccessor> getAllProcedureParameters(String callableName)
        throws SQLException {
    // 2 indentation (just a single 1 would align with the method code and won't stand out well)
```

or 
``` java
public Map<String, DbParameterAccessor> getAllProcedureParameters(
        String callableName) throws SQLException {

    // with multiple params
    public Map<String, DbParameterAccessor> getAllProcedureParameters(
            String callableName,
            String param2,
            String param3) throws SQLException, SQLException2 {
        // code
    }
```

At some point I used to do something like following but I think now I would rather take the approach in the previous sample.
``` java
public Map<String, DbParameterAccessor> getAllProcedureParameters(String callableName) throws SQLException {
```